gapis/replay/mapping_exporter.go

```diff
@@ -0,0 +1,158 @@
+// Copyright (C) 2019 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package replay
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"reflect"
+	"sort"
+
+	"github.com/google/gapid/core/data/binary"
+	"github.com/google/gapid/core/log"
+	"github.com/google/gapid/gapis/api"
+	"github.com/google/gapid/gapis/api/transform"
+	"github.com/google/gapid/gapis/memory"
+	"github.com/google/gapid/gapis/replay/builder"
+)
+
+type VulkanHandleMappingItem struct {
+	HandleType  string
+	TraceValue  string
+	ReplayValue string
+}
+
+type MappingExporter struct {
+	mappings *[]VulkanHandleMappingItem
+	thread   uint64
+	path     string
+}
+
+func NewMappingExporter(ctx context.Context, mappings *[]VulkanHandleMappingItem) *MappingExporter {
+	return &MappingExporter{
+		mappings: mappings,
+		thread:   0,
+		path:     "",
+	}
+}
+
+func NewMappingExporterWithPrint(ctx context.Context, path string) *MappingExporter {
+	mapping := make([]VulkanHandleMappingItem, 0, 0)
+	return &MappingExporter{
+		mappings: &mapping,
+		thread:   0,
+		path:     path,
+	}
+}
+
+func (m *MappingExporter) ExtractRemappings(ctx context.Context, s *api.GlobalState, b *builder.Builder, done func()) error {
+	var ret error
+
+	total := 0
+
+	for k, v := range b.Remappings {
+		typ := reflect.TypeOf(k)
+		var size uint64
+		if t, ok := k.(memory.SizedTy); ok {
+			size = t.TypeSize(s.MemoryLayout)
+		} else {
+			size = uint64(typ.Size())
+		}
+		if size != 1 && size != 2 && size != 4 && size != 8 {
+			// Ignore objects that are not handles
+			continue
+		}
+		// Count the number of actual Posts we expect
+		total += 1
+		func(k interface{}, typ reflect.Type, size uint64) {
+			b.Post(v, size, func(r binary.Reader, err error) {
+				defer func() {
+					total -= 1
+
+					if total == 0 {
+						done()
+					}
```

Comment:
I know this block isn't really your code, but I still think there's a very unlikely data race here.  Whenever we get a new PostData message, we spin up a new goroutine (https://github.com/google/gapid/blob/6aeb23ad58849ab89e564bf2c9ac5e44b69f5a60/gapis/replay/builder/builder.go#L810) which then calls this decrement and compare logic.  AFAIK we don't get to chose how goroutines get scheduled once they're in the pool without our own synchronization.  In the worst case, through some hiccup in grpc or go which leads to multiple goroutines in the pool, the running goroutine could be swapped out between the decrement and comparison.  Wrapping this in a `mutex.Lock()` and `defer mutex.Unlock()` should avoid any risk of a data race.

Reply:
I have moved it from the old mapping_printer. That will be rewritten with a different logic.